Memory access error when custom_specs folder doesn't exist

# Steps to reproduce﻿

1. create a `.enval.manifest.yaml` config using a non-existing path in the `custom_specs` field.
2. run `enval` using that manifest

# Expected behavior

Enval executes the checks using the default tool specs and shows a warning saying that the `custom_specs`folder doesn't exist.

# Actual behavior

Enval crashes with a moemory access error and showing a stack trace:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x13abddf]

goroutine 1 [running]:
github.com/Adhara-Tech/enval/pkg/infra.FileSystemToolsStorage.findInFileSystem.func1(0xc00012f268, 0x7, 0x0, 0x0, 0x15d7640, 0xc0001e60f0, 0x0, 0x14e3ea0)
        /Users/fllaca/projects/enval/pkg/infra/filesystem_tools_storage.go:29 +0x2f
path/filepath.Walk(0xc00012f268, 0x7, 0xc0001755a8, 0x15cc680, 0xc0001755d8)
        /usr/local/go/src/path/filepath/path.go:404 +0x6a
github.com/Adhara-Tech/enval/pkg/infra.FileSystemToolsStorage.findInFileSystem(0xc00012f268, 0x7, 0xc00012f298, 0x4, 0xc00015aaa0, 0x197, 0xc00015aa01)
        /Users/fllaca/projects/enval/pkg/infra/filesystem_tools_storage.go:28 +0x90
github.com/Adhara-Tech/enval/pkg/infra.FileSystemToolsStorage.Find(0xc00012f268, 0x7, 0xc00012f298, 0x4, 0x6, 0x0, 0x0)
        /Users/fllaca/projects/enval/pkg/infra/filesystem_tools_storage.go:62 +0x4d
github.com/Adhara-Tech/enval/pkg/infra.ToolsStorageChain.Find(0xc000123420, 0x2, 0x2, 0xc00012f298, 0x4, 0x0, 0x1b4, 0x14)
        /Users/fllaca/projects/enval/pkg/infra/chainned_tools_storage.go:27 +0x79
github.com/Adhara-Tech/enval/pkg/adapters.DefaultToolsStorageAdapter.Find(...)
        /Users/fllaca/projects/enval/pkg/adapters/tools_storage_adapter.go:39
github.com/Adhara-Tech/enval/pkg/manifestchecker.ToolsManager.ValidateTool(0x15d7060, 0xc000113dd0, 0x15db500, 0x1935488, 0xc0001e60c0, 0xc00012f298, 0x4, 0x0, 0x0, 0x0, ...)
        /Users/fllaca/projects/enval/pkg/manifestchecker/tool_checker_manager.go:67 +0x66
github.com/Adhara-Tech/enval/pkg/manifestchecker.ToolsManager.ValidateManifestAndNotify(0x15d7060, 0xc000113dd0, 0x15db500, 0x1935488, 0xc0001e60c0, 0xc000130c20, 0x1c, 0xc00012f268, 0x7, 0xc000152300, ...)
        /Users/fllaca/projects/enval/pkg/manifestchecker/tool_checker_manager.go:44 +0x112
main.executeCmd(0x1902580, 0x1935488, 0x0, 0x0, 0x0, 0x0)
        /Users/fllaca/projects/enval/cmd/enval.go:93 +0x46e
github.com/spf13/cobra.(*Command).execute(0x1902580, 0xc0001121e0, 0x0, 0x0, 0x1902580, 0xc0001121e0)
        /Users/fllaca/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0x1902580, 0xc000175f60, 0x1, 0x1)
        /Users/fllaca/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/fllaca/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
        /Users/fllaca/projects/enval/cmd/enval.go:49 +0x1ec

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,14 @@ For each changelong entry sections are:
 
 ## [Unreleased]
 
+### Added
+
+- Custom EnvalError implementation supporting error kinds [#45](https://github.com/Adhara-Tech/enval/issues/45)
+
+### Fixed
+
+- Memory access error when custom_specs folder doesn't exist [#49](https://github.com/Adhara-Tech/enval/issues/49)
+- Stacktrace is shown when a spec wasn't found [#34](https://github.com/Adhara-Tech/enval/issues/34)
 
 ## [0.1.0-rc2] - 2020-05-10
 

--- a/cmd/enval.go
+++ b/cmd/enval.go
@@ -47,7 +47,7 @@ func main() {
 
 	viper.AutomaticEnv()
 	if err := cmd.Execute(); err != nil {
-		fmt.Println(exerrors.ErrorStack(err))
+		fmt.Println(exerrors.PrintError(err))
 		os.Exit(1)
 	}
 }
@@ -90,9 +90,16 @@ func executeCmd(_ *cobra.Command, _ []string) error {
 	versionCheckerManager := manifestchecker.NewVersionCheckerManager(fieldVersionValidatorManager)
 	toolsManager := manifestchecker.NewToolsManager(toolsStorageAdapter, systemAdapter, versionCheckerManager)
 
+	_, err = toolsManager.IsManifestCompliant(*manifest)
+	if err != nil {
+		prettyPrintError(err)
+		os.Exit(1)
+	}
+
 	_, err = toolsManager.ValidateManifestAndNotify(*manifest, cmdNotifier)
 	if err != nil {
-		return err
+		prettyPrintError(err)
+		os.Exit(1)
 	}
 
 	return nil
@@ -107,6 +114,10 @@ func toolName(tool model.ManifestTool) string {
 		return fmt.Sprintf("%s(%s)", tool.Name, *tool.Flavor)
 	}
 	return tool.Name
+}
+
+func prettyPrintError(err error) {
+	fmt.Printf("%s%s%s Invalid Manifest:\n\t%s", invalidSymbol, invalidSymbol, invalidSymbol, exerrors.PrintError(err))
 }
 
 func renderVersions(tool model.ManifestTool, fieldVersions map[string]manifestchecker.FieldValidationResult) string {

--- a/pkg/adapters/system_adapter.go
+++ b/pkg/adapters/system_adapter.go
@@ -1,6 +1,7 @@
 package adapters
 
 import (
+	"os"
 	"os/exec"
 
 	"github.com/Adhara-Tech/enval/pkg/exerrors"
@@ -23,7 +24,7 @@ func (systemAdapter DefaultSystemAdapter) CheckCommandAvailable(command string) 
 				return false, nil
 			}
 		}
-		return false, exerrors.Wrap(err)
+		return false, exerrors.Wrap(err, exerrors.InternalEnvalErrorKind)
 	}
 
 	return true, nil
@@ -34,10 +35,27 @@ func (systemAdapter DefaultSystemAdapter) ExecuteCommand(commandName string, par
 
 	versionString, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", exerrors.Wrap(err)
+		return "", exerrors.Wrap(err, exerrors.InternalEnvalErrorKind)
 	}
 
 	return string(versionString), nil
+}
+
+func (systemAdapter DefaultSystemAdapter) CheckDirExist(path string) (bool, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+
+		return false, exerrors.Wrap(err, exerrors.InternalEnvalErrorKind)
+	}
+
+	if !fileInfo.IsDir() {
+		return false, nil
+	}
+
+	return true, nil
 }
 
 var _ manifestchecker.SystemAdapter = (*DefaultSystemAdapter)(nil)

--- a/pkg/adapters/tools_storage_adapter.go
+++ b/pkg/adapters/tools_storage_adapter.go
@@ -2,7 +2,6 @@ package adapters
 
 import (
 	"fmt"
-	"strings"
 
 	"github.com/Adhara-Tech/enval/pkg/exerrors"
 	"github.com/Adhara-Tech/enval/pkg/manifestchecker"
@@ -27,12 +26,11 @@ type ToolsStorage interface {
 }
 
 func NewToolNotFoundExError(toolName string) error {
-	return exerrors.New(fmt.Sprintf("tool with name [%s] not found", toolName))
+	return exerrors.New(fmt.Sprintf("tool with name [%s] not found", toolName), exerrors.ToolDefinitionNotFoundEnvalErrorKind)
 }
 
 func IsToolNotFoundExError(err error) bool {
-	errStack := exerrors.ErrorStack(err)
-	return strings.HasPrefix(errStack, "tool with name")
+	return exerrors.IsEnvalErrorWithKind(err, exerrors.ToolDefinitionNotFoundEnvalErrorKind)
 }
 
 func (adapter DefaultToolsStorageAdapter) Find(toolName string) (*manifestchecker.ToolSpec, error) {

--- a/pkg/exerrors/extended_error.go
+++ b/pkg/exerrors/extended_error.go
@@ -6,19 +6,60 @@ import (
 	goerrors "github.com/go-errors/errors"
 )
 
-func Wrap(err error) error {
-	return goerrors.Wrap(err, 3)
+func Wrap(err error, kind EnvalErrorKind) error {
+	return EnvalError{
+		innerError: goerrors.Wrap(err, 3),
+		kind:       kind,
+	}
 }
 
-func New(msg string) error {
-	return goerrors.Errorf(msg)
+func New(msg string, kind EnvalErrorKind) error {
+	return EnvalError{
+		innerError: goerrors.Errorf(msg),
+		kind:       kind,
+	}
 }
 
-func ErrorStack(err error) string {
-	goError, ok := err.(*goerrors.Error)
+func PrintError(err error) string {
+	envalError, ok := err.(EnvalError)
 	if ok {
-		return fmt.Sprintf("%s\n%s", goError.Error(), goError.ErrorStack())
+		return envalError.Error()
 	} else {
-		return err.Error()
+		wrappedError := Wrap(err, InternalEnvalErrorKind)
+		return wrappedError.Error()
+	}
+}
+
+type EnvalErrorKind int
+
+const (
+	InternalEnvalErrorKind = iota
+	ToolDefinitionNotFoundEnvalErrorKind
+	FlavorDefinitionNotFoundEnvalErrorKind
+	UnknownParserEnvalErrorKind
+	UnsupportedInputRawVersionEnvalErrorKind
+	FieldVersionKeyNotFoundEnvalErrorKind
+	InvalidCustomSpecsDirEnvalErrorKind
+)
+
+func IsEnvalErrorWithKind(err error, kind EnvalErrorKind) bool {
+	envalError, ok := err.(EnvalError)
+	if !ok {
+		return false
+	}
+
+	return envalError.kind == kind
+}
+
+type EnvalError struct {
+	innerError *goerrors.Error
+	kind       EnvalErrorKind
+}
+
+func (err EnvalError) Error() string {
+	if err.kind == InternalEnvalErrorKind {
+		return fmt.Sprintf("%s\n%s", err.innerError.Error(), err.innerError.ErrorStack())
+	} else {
+		return err.innerError.Error()
 	}
 }

--- a/pkg/infra/boxed_tools_storage.go
+++ b/pkg/infra/boxed_tools_storage.go
@@ -34,7 +34,7 @@ func findInBox(box ListFinder, toolsFindOptions adapters.ToolFindOptions) (*mani
 	for _, currentToolSpec := range box.List() {
 		toolSpecBytes, err := box.Find(currentToolSpec)
 		if err != nil {
-			return nil, exerrors.Wrap(err)
+			return nil, exerrors.Wrap(err, exerrors.InternalEnvalErrorKind)
 		}
 
 		tool := &manifestchecker.ToolSpec{}

--- a/pkg/infra/filesystem_tools_storage.go
+++ b/pkg/infra/filesystem_tools_storage.go
@@ -34,13 +34,13 @@ func (storage FileSystemToolsStorage) findInFileSystem(toolsFindOptions adapters
 	})
 
 	if err != nil {
-		return nil, exerrors.Wrap(err)
+		return nil, exerrors.Wrap(err, exerrors.InternalEnvalErrorKind)
 	}
 
 	for _, currentToolSpec := range toolSpecFiles {
 		toolSpecBytes, err := ioutil.ReadFile(currentToolSpec)
 		if err != nil {
-			return nil, exerrors.Wrap(err)
+			return nil, exerrors.Wrap(err, exerrors.InternalEnvalErrorKind)
 		}
 
 		tool := &manifestchecker.ToolSpec{}

--- a/pkg/manifestchecker/test_helpers.go
+++ b/pkg/manifestchecker/test_helpers.go
@@ -11,6 +11,10 @@ type TestSystemAdapter struct {
 	outputFilePath string
 }
 
+func (t *TestSystemAdapter) CheckDirExist(path string) (bool, error) {
+	panic("implement me")
+}
+
 func (t *TestSystemAdapter) NextOutput(outputFilePath string) {
 	t.outputFilePath = outputFilePath
 }

--- a/pkg/manifestchecker/tool_checker_manager.go
+++ b/pkg/manifestchecker/tool_checker_manager.go
@@ -1,8 +1,9 @@
 package manifestchecker
 
 import (
-	"github.com/Adhara-Tech/enval/pkg/exerrors"
+	"fmt"
 
+	"github.com/Adhara-Tech/enval/pkg/exerrors"
 	"github.com/Adhara-Tech/enval/pkg/model"
 )
 
@@ -14,6 +15,8 @@ type ToolsManager struct {
 
 type SystemAdapter interface {
 	CheckCommandAvailable(command string) (bool, error)
+	// Returns true if the directory exists. Returns false if the path does not exist or if it is a file
+	CheckDirExist(path string) (bool, error)
 	ExecuteCommand(commandName string, params []string) (string, error)
 }
 
@@ -50,6 +53,29 @@ func (tm ToolsManager) ValidateManifestAndNotify(manifest model.Manifest, notify
 	}
 
 	return result, nil
+}
+
+func (tm ToolsManager) IsManifestCompliant(manifest model.Manifest) (bool, error) {
+	if manifest.CustomSpecs != "" {
+		ok, err := tm.systemAdapter.CheckDirExist(manifest.CustomSpecs)
+		if err != nil {
+			return false, err
+		}
+
+		if !ok {
+			return false, exerrors.New(fmt.Sprintf("Custom specs directory [%s] defined in manifest does not exist or it is not a directory", manifest.CustomSpecs), exerrors.InvalidCustomSpecsDirEnvalErrorKind)
+		}
+
+	}
+
+	for _, manifestTool := range manifest.Tools {
+		_, err := tm.toolsStorageAdapter.Find(manifestTool.Name)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
 }
 
 // * Manifest has flavor configured
@@ -152,7 +178,7 @@ func (tm ToolsManager) executeVersionCommand(tool ToolSpec, flavor *string) (*ex
 
 	commandAvailable, err := tm.systemAdapter.CheckCommandAvailable(tool.Command)
 	if err != nil {
-		return nil, exerrors.Wrap(err)
+		return nil, err
 	}
 
 	if !commandAvailable {

--- a/pkg/manifestchecker/tool_checker_manager_test.go
+++ b/pkg/manifestchecker/tool_checker_manager_test.go
@@ -90,17 +90,17 @@ func TestToolsManager_Validate(t *testing.T) {
 			Manifest:              manifestchecker.ManifestFrom("python", map[string]string{"version": "= 3.7.3"}),
 			VersionOutputFilePath: "testdata/tools-version-output/python_3.7.3.output.txt",
 		},
-    {
+		{
 			TestName:              "ruby 2.6.3",
 			Manifest:              manifestchecker.ManifestFrom("ruby", map[string]string{"version": "= 2.6.3"}),
 			VersionOutputFilePath: "testdata/tools-version-output/ruby_2.6.3.output.txt",
 		},
-    {
+		{
 			TestName:              "elixir 1.10.2",
 			Manifest:              manifestchecker.ManifestFrom("elixir", map[string]string{"version": "= 1.10.2"}),
 			VersionOutputFilePath: "testdata/tools-version-output/elixir_1.10.2.output.txt",
 		},
-    {
+		{
 			TestName:              "phoenix 1.5.1",
 			Manifest:              manifestchecker.ManifestFrom("phoenix", map[string]string{"version": "= 1.5.1"}),
 			VersionOutputFilePath: "testdata/tools-version-output/phoenix_1.5.1.output.txt",
@@ -125,7 +125,7 @@ func TestToolsManager_Validate(t *testing.T) {
 
 			result, err := toolsCheckerManager.ValidateManifest(currentTestData.Manifest)
 			if err != nil {
-				fmt.Println(exerrors.ErrorStack(err))
+				fmt.Println(exerrors.PrintError(err))
 			}
 			require.Nil(t, err)
 			require.NotNil(t, result)

--- a/pkg/manifestchecker/types.go
+++ b/pkg/manifestchecker/types.go
@@ -29,7 +29,7 @@ func (t ToolSpec) findFlavor(flavor string) (*FlavorSpec, error) {
 		}
 	}
 
-	return nil, exerrors.New(fmt.Sprintf("flavor [%s] not found for tool [%s]", flavor, t.Name))
+	return nil, exerrors.New(fmt.Sprintf("flavor [%s] not found for tool [%s]", flavor, t.Name), exerrors.FlavorDefinitionNotFoundEnvalErrorKind)
 }
 
 func (t ToolSpec) consolidateVersionCommandArgsForFlavor(flavor *string) ([]string, error) {

--- a/pkg/manifestchecker/version_checker_manager.go
+++ b/pkg/manifestchecker/version_checker_manager.go
@@ -46,7 +46,7 @@ func (versionCheckerManager VersionCheckerManager) buildVersionChecker(checkVers
 			regexpVersionParser := NewRegexVersionParser(parser.Regexp, versionCheckerSpec.FieldNames())
 			versionParsers[index] = regexpVersionParser
 		} else {
-			return nil, exerrors.New(fmt.Sprintf("unknown parser type [%s]", parser.Type))
+			return nil, exerrors.New(fmt.Sprintf("unknown parser type [%s]", parser.Type), exerrors.UnknownParserEnvalErrorKind)
 		}
 	}
 

--- a/pkg/manifestchecker/version_parser.go
+++ b/pkg/manifestchecker/version_parser.go
@@ -2,7 +2,6 @@ package manifestchecker
 
 import (
 	"fmt"
-	"strings"
 
 	"github.com/Adhara-Tech/enval/pkg/exerrors"
 )
@@ -16,14 +15,12 @@ type VersionParser interface {
 	Parse(rawVersion string) (map[string]string, error)
 }
 
-const unsupportedInputRawVersion = "unsupported input raw version"
-
 // NewUnsupportedInputRawVersionError creates a new UnsupportedInputRawVersionError
 func NewUnsupportedInputRawVersionError(msg string) error {
-	return exerrors.New(fmt.Sprint(unsupportedInputRawVersion, ". ", msg))
+	return exerrors.New(fmt.Sprint("unsupported input raw version. ", msg), exerrors.UnsupportedInputRawVersionEnvalErrorKind)
 }
 
 // IsUnsupportedInputRawVersionError checks if the given error is of type UnsupportedInputRawVersionError
 func IsUnsupportedInputRawVersionError(err error) bool {
-	return strings.HasPrefix(err.Error(), unsupportedInputRawVersion)
+	return exerrors.IsEnvalErrorWithKind(err, exerrors.UnsupportedInputRawVersionEnvalErrorKind)
 }

--- a/pkg/manifestchecker/version_parser_regexp.go
+++ b/pkg/manifestchecker/version_parser_regexp.go
@@ -50,7 +50,7 @@ func (parser *RegexpVersionParser) Parse(rawVersion string) (map[string]string, 
 
 	for _, currentKey := range parser.keys {
 		if _, ok := resultMap[currentKey]; !ok {
-			return nil, exerrors.New(fmt.Sprintf("key [%s] not found as part of regexp to find version fields on version string [%s]", currentKey, rawVersion))
+			return nil, exerrors.New(fmt.Sprintf("key [%s] not found as part of regexp to find version fields on version string [%s]", currentKey, rawVersion), exerrors.FieldVersionKeyNotFoundEnvalErrorKind)
 		}
 
 	}

